src/vmm_unix.ml

```diff
@@ -486,12 +495,111 @@ let dump_block name =
+  let open Lwt.Infix in
+  let fd = Lwt_unix.of_unix_file_descr fd in
+  Lwt.catch (fun () ->
+      let rec more fd size stream off =
+        let len = Int.min (size - off) 4096 in
+        let buf = Bytes.create len in
+        Lwt_unix.read fd buf 0 len >>= fun read ->
+        let buf = if read = len then buf else Bytes.sub buf 0 read in
+        stream#push (Bytes.unsafe_to_string buf) >>= fun () ->
+        if read = size - off then begin
+          stream#close;
+          Lwt.return_unit
+        end else
+          more fd size stream (off + read)
+      in
+      more fd size stream 0 >>= fun () ->
+      safe_close fd >|= fun () ->
+      Ok ())
+    (function
+      | Lwt.Canceled ->
+        (* We assume error reporting is done by the canceller *)
+        Lwt.return (Ok ())
+      | e ->
+       Logs.err (fun m -> m "error streaming %a: %s" Fpath.pp name
+                    (Printexc.to_string e));
+       Lwt.return (Error (`Msg "streaming block device")))
+
+let open_block_fd name =
+  let block_name = block_file name in
+  let* exists = Bos.OS.File.exists block_name in
+  if not exists then
+    Error (`Msg "file does not exist")
+  else
+    let name_str = Fpath.to_string block_name in
+    let size = Unix.(stat name_str).st_size in
+    let fd = openfile name_str [ O_RDONLY ] 0 in
+    Ok (fd, size, block_name)
+
 let mb_of_bytes size =
   if size = 0 || size land 0xFFFFF <> 0 then
     Error (`Msg "size is either 0 or not MB aligned")
   else
     Ok (size lsr 20)
 
+let stream_to_fd ?(byte_size = Int.max_int) fd stream name =
+  let open Lwt.Infix in
+  let exception Malformed_data of string in
+  Lwt.catch (fun () ->
+      let rec read_more fd stream size =
+        Lwt_stream.get stream >>= function
+        | None -> Lwt.return (Ok ())
+        | Some `Data data ->
+          if String.length data > byte_size - size then begin
+            Logs.err (fun m -> m "stream exceeds size");
+            Lwt.return (Error (`Msg "stream exceeds size"))
+          end else
+            let rec write fd data off len =
+              let to_write = len - off in
+              Lwt_unix.write fd data off to_write >>= fun written ->
+              if written = to_write then
+                read_more fd stream (size + len)
+              else
+                write fd data (off + written) len
+            in
+            write fd (Bytes.unsafe_of_string data) 0 (String.length data)
+        | Some `Malformed msg ->
+          raise (Malformed_data msg)
+      in
+      read_more fd stream 0 >>= fun r ->
+      safe_close fd >|= fun () ->
+      r)
+    (fun e ->
+       Logs.err (fun m -> m "error writing %a: %s"
+                    Fpath.pp name (Printexc.to_string e));
+       safe_close fd >>= fun () ->
+       let rec drop_stream s =
+         Lwt_stream.get s >>= function
+         | None -> Lwt.return_unit
+         | Some _ -> drop_stream s
+       in
+       drop_stream stream >|= fun () ->
```

Comment:
Here we should cancel the reader I think :/

Reply:
Or maybe it's fine - the other end will not expect a message until they've sent all their data.

Reply:
Ahhh hmm I'm now not sure this works if the compressed data is malformed. In that case we close the uncompressed stream, and then we can't consume the remainder of the underlying compressed stream.